rust/primary/src/core.rs

```diff
@@ -0,0 +1,411 @@
+    processing: HashMap<Round, HashSet<Digest>>,
+    /// The last header we proposed (for which we are waiting votes).
+    current_header: Header,
+    /// Aggregates votes into a certificate.
+    votes_aggregator: VotesAggregator,
+    /// Aggregates certificates to use as parents for new headers.
+    certificates_aggregators: HashMap<Round, Box<CertificatesAggregator>>,
+    /// A network sender to send the batches to the other workers.
+    network: ReliableSender,
+    /// Keeps the cancel handlers of the messages we sent.
+    cancel_handlers: HashMap<Round, Vec<CancelHandler>>,
+}
+
+impl Core {
+    #[allow(clippy::too_many_arguments)]
+    pub fn spawn(
+        name: PublicKey,
+        committee: Committee,
+        store: Store,
+        synchronizer: Synchronizer,
+        signature_service: SignatureService,
+        consensus_round: Arc<AtomicU64>,
+        gc_depth: Round,
+        rx_primaries: Receiver<PrimaryMessage>,
+        rx_header_waiter: Receiver<Header>,
+        rx_certificate_waiter: Receiver<Certificate>,
+        rx_proposer: Receiver<Header>,
+        tx_consensus: Sender<Certificate>,
+        tx_proposer: Sender<(Vec<Digest>, Round)>,
+    ) {
+        tokio::spawn(async move {
+            Self {
+                name,
+                committee,
+                store,
+                synchronizer,
+                signature_service,
+                consensus_round,
+                gc_depth,
+                rx_primaries,
+                rx_header_waiter,
+                rx_certificate_waiter,
+                rx_proposer,
+                tx_consensus,
+                tx_proposer,
+                gc_round: 0,
+                last_voted: HashMap::with_capacity(2 * gc_depth as usize),
+                processing: HashMap::with_capacity(2 * gc_depth as usize),
+                current_header: Header::default(),
+                votes_aggregator: VotesAggregator::new(),
+                certificates_aggregators: HashMap::with_capacity(2 * gc_depth as usize),
+                network: ReliableSender::new(),
+                cancel_handlers: HashMap::with_capacity(2 * gc_depth as usize),
+            }
+            .run()
+            .await;
+        });
+    }
+
+    async fn process_own_header(&mut self, header: Header) -> DagResult<()> {
+        // Reset the votes aggregator.
+        self.current_header = header.clone();
+        self.votes_aggregator = VotesAggregator::new();
+
+        // Broadcast the new header in a reliable manner.
+        let addresses = self
+            .committee
+            .others_primaries(&self.name)
+            .iter()
+            .map(|(_, x)| x.primary_to_primary)
+            .collect();
+        let bytes = bincode::serialize(&PrimaryMessage::Header(header.clone()))
+            .expect("Failed to serialize our own header");
+        let handlers = self.network.broadcast(addresses, Bytes::from(bytes)).await;
```

Comment:
do we send the header without our signature on? we assume the connections are secure and no MITM?